File: app/src/main/java/com/cooper/wheellog/ScanActivity.java
package com.cooper.wheellog;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.provider.Settings;
import android.text.InputType;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.cooper.wheellog.utils.SettingsUtil;

import java.util.Arrays;

import timber.log.Timber;

public class ScanActivity extends AppCompatActivity {
    private DeviceListAdapter mDeviceListAdapter;
    private BluetoothAdapter mBluetoothAdapter;
    private boolean mScanning;
    private Handler mHandler = new Handler();
    private ProgressBar pb;
    private TextView scanTitle;

    // Stops scanning after 10 seconds.
    private static final long SCAN_PERIOD = 10000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final ViewGroup nullParent = null;
        View convertView = getLayoutInflater().inflate(R.layout.activity_scan, nullParent);
        ListView lv = (ListView) convertView.findViewById(android.R.id.list);
        pb = (ProgressBar) convertView.findViewById(R.id.scanProgress);
        scanTitle = (TextView) convertView.findViewById(R.id.scan_title);
        lv.setOnItemClickListener(onItemClickListener);
        mDeviceListAdapter = new DeviceListAdapter(this);
        lv.setAdapter(mDeviceListAdapter);
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(this, R.style.AppTheme_Dialog_Alert)
                .setView(convertView)
                .setCancelable(false)
                .setOnKeyListener(new DialogInterface.OnKeyListener() {
                    @Override
                    public boolean onKey(DialogInterface dialogInterface, int keycode, KeyEvent keyEvent) {
                        if (keycode == KeyEvent.KEYCODE_BACK &&
                                keyEvent.getAction() == KeyEvent.ACTION_UP &&
                                !keyEvent.isCanceled()) {
                            if (mScanning)
                                scanLeDevice(false);
                            mHandler.removeCallbacksAndMessages(null);
                            dialogInterface.cancel();
                            finish();
                            return true;
                        }
                        return false;
                    }
                });
        alertDialog.show();

        if (!isLocationEnabled(this)) {
            Intent myIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
            startActivity(myIntent);
        }

        // Initializes a Bluetooth adapter.  For API level 18 and above, get a reference to
        // BluetoothAdapter through BluetoothManager.
        final BluetoothManager bluetoothManager =
                (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
        mBluetoothAdapter = bluetoothManager.getAdapter();
    }

    @Override
    protected void onResume() {
        super.onResume();

        scanLeDevice(true);
    }

    private AdapterView.OnItemClickListener onItemClickListener = new AdapterView.OnItemClickListener() {

        @Override
        public void onItemClick(AdapterView<?> adapterView, final View view, int i, long l) {
            if (mScanning)
                scanLeDevice(false);
            mHandler.removeCallbacksAndMessages(null);
            final String deviceAddress = mDeviceListAdapter.getDevice(i).getAddress();
            final String deviceName = mDeviceListAdapter.getDevice(i).getName();
            final String advData = mDeviceListAdapter.getAdvData(i);
            Timber.i("Device selected MAC = %s", deviceAddress);
            Timber.i("Device selected Name = %s", deviceName);
            Timber.i("Device selected Data = %s", advData);
            Intent intent = new Intent();
            intent.putExtra("MAC", deviceAddress);
            intent.putExtra("NAME", deviceName);
            intent.putExtra("ADV", advData);
            SettingsUtil.setAdvDataForWheel(view.getContext(), deviceAddress, advData);
            setResult(RESULT_OK, intent);
            //Ask for inmotion password
            AlertDialog.Builder builder = new AlertDialog.Builder(view.getContext());
            builder.setTitle(R.string.wheel_pass_imotion);

            final EditText input = new EditText(view.getContext());
            input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
            builder.setView(input);
            builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    String password = input.getText().toString();
                    SettingsUtil.setPasswordForWheel(view.getContext(), deviceAddress, password);
                    finish();
                }
            });
            builder.setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.cancel();
                    finish();
                }
            });
            builder.show();
        }
    };

    // Device scan callback.
    private BluetoothAdapter.LeScanCallback mLeScanCallback =
            new BluetoothAdapter.LeScanCallback() {

                @Override
                public void onLeScan(final BluetoothDevice device, int rssi, byte[] scanRecord) {
                    final String manufacturerData = findManufacturerData(scanRecord); // 4e421300000000ec
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mDeviceListAdapter.addDevice(device, manufacturerData);
                            mDeviceListAdapter.notifyDataSetChanged();
                        }
                    });
                }
            };


    public String findManufacturerData(byte[] scanRecord) {
        int index = 0;
        String result = "";
        while (index < scanRecord.length) {
            int length = scanRecord[index++];
            //Done once we run out of records
            if (length == 0) break;

            int type = scanRecord[index];
            //Done if our record isn't a valid type
            if (type == 0) break;

            byte[] data = Arrays.copyOfRange(scanRecord, index+1, index+length);

            //Advance
            index += length;
            if (type == -1) {
                result = ByteArrayToString(data);
            }
        }
        Timber.i("Found data: " + result);
        return result;
    }

    public static String ByteArrayToString(byte[] ba)
    {
        String hexString = "";
        for(int i = 0; i < ba.length; i++){
            String thisByte = "".format("%02x", ba[i]);
            hexString += thisByte;
        }
        return hexString;
    }


    private void scanLeDevice(final boolean enable) {
        if (enable) {
            // Stops scanning after a pre-defined scan period.
            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    scanLeDevice(false);
                }
            }, SCAN_PERIOD);

            mScanning = true;
            mBluetoothAdapter.startLeScan(mLeScanCallback);
            pb.setVisibility(View.VISIBLE);
            scanTitle.setText(R.string.scanning);
        } else {
            mScanning = false;
            mBluetoothAdapter.stopLeScan(mLeScanCallback);
            pb.setVisibility(View.GONE);
            scanTitle.setText(R.string.devices);
        }
    }

    public static boolean isLocationEnabled(Context context) {
        int locationMode = 0;
        String locationProviders;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT){
            try {
                locationMode = Settings.Secure.getInt(context.getContentResolver(), Settings.Secure.LOCATION_MODE);

            } catch (Settings.SettingNotFoundException e) {
                e.printStackTrace();
            }

            return locationMode != Settings.Secure.LOCATION_MODE_OFF;

        }else{
            locationProviders = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.LOCATION_PROVIDERS_ALLOWED);
            return !TextUtils.isEmpty(locationProviders);
        }
    }
}

File: app/src/main/java/com/cooper/wheellog/ScanActivity.kt
package com.cooper.wheellog

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.le.ScanResult
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.provider.Settings.SettingNotFoundException
import android.view.Gravity
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.widget.AdapterView.OnItemClickListener
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.cooper.wheellog.databinding.ActivityScanBinding
import com.cooper.wheellog.utils.PermissionsUtil
import com.cooper.wheellog.utils.StringUtil
import com.cooper.wheellog.utils.StringUtil.toHexStringRaw
import com.welie.blessed.BluetoothCentralManager
import com.welie.blessed.BluetoothCentralManagerCallback
import com.welie.blessed.BluetoothPeripheral
import com.welie.blessed.Transport
import timber.log.Timber


class ScanActivity: AppCompatActivity() {
    private var mDeviceListAdapter: DeviceListAdapter? = null
    private val central: BluetoothCentralManager by lazy {
        BluetoothCentralManager(
            this,
            bluetoothCentralManagerCallback,
            Handler(Looper.getMainLooper())
        ).apply { transport = Transport.LE }
    }
    private var pb: ProgressBar? = null
    private var scanTitle: TextView? = null
    // Stops scanning after 10 seconds.
    private val scanPeriodHandler = Handler(Looper.getMainLooper())
    private val scanPeriod: Long = 10_000
    private lateinit var alertDialog: AlertDialog
    private lateinit var macLayout: LinearLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityScanBinding.inflate(layoutInflater, null, false)
        pb = binding.scanProgress
        scanTitle = binding.scanTitle
        mDeviceListAdapter = DeviceListAdapter(this)
        binding.list.onItemClickListener = onItemClickListener
        binding.list.adapter = mDeviceListAdapter
        macLayout = binding.lastMacText
        binding.lastMacText.editText!!.setText(WheelLog.AppConfig.lastMac)
        binding.lastMacText.setEndIconOnClickListener {
            val deviceAddress = binding.lastMacText.editText?.text.toString()
            if (!StringUtil.isCorrectMac(deviceAddress)) {
                binding.lastMacText.error = "incorrect MAC"
                binding.lastMacText.errorIconDrawable = null
                return@setEndIconOnClickListener
            }
            if (central.isScanning) {
                scanLeDevice(false)
            }
            val intent = Intent()
            intent.putExtra("MAC", deviceAddress)
            WheelLog.AppConfig.lastMac = deviceAddress
            setResult(RESULT_OK, intent)
            WheelLog.AppConfig.passwordForWheel = ""
            close()
        }
        alertDialog = AlertDialog.Builder(this, R.style.OriginalTheme_Dialog_Alert)
                .setView(binding.root)
                .setCancelable(false)
                .setOnKeyListener { dialogInterface: DialogInterface, keycode: Int, keyEvent: KeyEvent ->
                    if (keycode == KeyEvent.KEYCODE_BACK && keyEvent.action == KeyEvent.ACTION_UP &&
                            !keyEvent.isCanceled) {
                        if (central.isScanning) {
                            scanLeDevice(false)
                        }
                        dialogInterface.cancel()
                        close()
                    }
                    false
                }
                .create()
        window.attributes = alertDialog.window?.attributes?.apply {
            gravity = Gravity.TOP
            flags = flags and WindowManager.LayoutParams.FLAG_DIM_BEHIND.inv()
        }
        alertDialog.show()
        if (!isLocationEnabled(this)) {
            val myIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
            startActivity(myIntent)
        }
    }

    private val bluetoothCentralManagerCallback: BluetoothCentralManagerCallback =
        object : BluetoothCentralManagerCallback() {
            override fun onDiscoveredPeripheral(
                peripheral: BluetoothPeripheral,
                scanResult: ScanResult
            ) {
                super.onDiscoveredPeripheral(peripheral, scanResult)
                val scanRecord = scanResult.scanRecord?.bytes ?: byteArrayOf()
                if (scanResult.scanRecord != null) {
                    val manufacturerData = findManufacturerData(scanRecord) // 4e421300000000ec
                    runOnUiThread {
                        mDeviceListAdapter!!.addDevice(scanResult.device, manufacturerData)
                        mDeviceListAdapter!!.notifyDataSetChanged()
                    }
                }
            }
        }

    private fun close () {
        central.stopScan()
        alertDialog.dismiss()
        finish()
    }

    override fun onResume() {
        super.onResume()
        if (central.isBluetoothEnabled) {
            if (!PermissionsUtil.checkBlePermissions(this)) {
                if (PermissionsUtil.isMaxBleReq) {
                    killMe()
                }
                return
            }
            scanLeDevice(true)
        } else {
            if (PermissionsUtil.checkBlePermissions(this)) {
                val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
                ActivityCompat.startActivityForResult(this, enableBtIntent, 2, null)
            } else {
                killMe()
            }
        }
    }

    private fun killMe() {
        central.close()
        alertDialog.dismiss()
        finish()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.all {r -> r == PackageManager.PERMISSION_GRANTED}) {
            scanLeDevice(true)
        }
    }

    @SuppressLint("MissingPermission")
    private val onItemClickListener = OnItemClickListener { _, _, i, _ ->
        if (central.isScanning) {
            central.stopScan()
        }
        val device = mDeviceListAdapter!!.getDevice(i)
        val deviceAddress = device.address
        val deviceName = device.name
        val advData = mDeviceListAdapter!!.getAdvData(i)
        Timber.i("Device selected MAC = %s", deviceAddress)
        Timber.i("Device selected Name = %s", deviceName)
        Timber.i("Device selected Data = %s", advData)
        val intent = Intent()
        intent.putExtra("MAC", deviceAddress)
        intent.putExtra("NAME", deviceName)
        WheelLog.AppConfig.lastMac = deviceAddress
        WheelLog.AppConfig.advDataForWheel = advData
        setResult(RESULT_OK, intent)
        // Set password for inmotion
        WheelLog.AppConfig.passwordForWheel = ""
        close()
    }

    private infix fun Byte.eq(i: Int): Boolean = this == i.toByte()

    private fun findManufacturerData(scanRecord: ByteArray): String {
        var index = 0
        var result = ""
        while (index < scanRecord.size) {
            val length = scanRecord[index++]
            // Done once we run out of records
            val toIndex = index + length
            if (length eq 0 || toIndex > scanRecord.size) {
                break
            }
            val type = scanRecord[index]
            // Done if our record isn't a valid type
            if (type eq 0) {
                break
            }
            val data = scanRecord.copyOfRange(index + 1, toIndex)

            // Advance
            index = toIndex
            if (type eq -1) {
                result = toHexStringRaw(data)
            }
        }
        Timber.i("Found data: %s", result)
        return result
    }

    private fun scanLeDevice(enable: Boolean) {
        if (enable) {
            // Stops scanning after a pre-defined scan period.
            scanPeriodHandler.postDelayed({ scanLeDevice(false) }, scanPeriod)
            central.stopScan()
            central.scanForPeripherals()
            pb!!.visibility = View.VISIBLE
            scanTitle!!.setText(R.string.scanning)
            macLayout.visibility = View.GONE
        } else {
            scanPeriodHandler.removeCallbacksAndMessages(null)
            central.stopScan()
            pb!!.visibility = View.GONE
            scanTitle!!.setText(R.string.devices)
            macLayout.visibility = View.VISIBLE
        }
    }

    private fun isLocationEnabled(context: Context): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            // This is new method provided in API 28
            val lm = context.getSystemService(LOCATION_SERVICE) as LocationManager
            lm.isLocationEnabled
        } else {
            // This is Deprecated in API 28
            val locationMode = try {
                Settings.Secure.getInt(context.contentResolver, Settings.Secure.LOCATION_MODE)
            } catch (e: SettingNotFoundException) {
                e.printStackTrace()
            }
            locationMode != Settings.Secure.LOCATION_MODE_OFF
        }
    }
}